### PR TITLE
Add support to decode a BGP Shutdown Communication

### PR DESCRIPTION
**Pull request by job**
The draft-ietf-idr-shutdown-02 document specifies a mechanism to transmit a short freeform UTF-8 [RFC3629] message as part of a Cease NOTIFICATION message [RFC4486] to inform the peer why the BGP session is being shutdown.

### CODE REVIEW

**Comment by infrastation** [comment]
Does the test pass in your working copy?

**Comment by job** [comment]
@infrastation yes

**Comment by job** [comment]
@infrastation seems it works on macosx, ubuntu, but not on arch linux and whatever travis is using

**Comment by infrastation** [comment]
Such behaviour often hints at an alignment issue. Would be useful to have a way to reproduce the test failure and look at the backtrace to see where the issue is.

File: print-bgp.c [hunk 1]
```diff
@@ -2669,7 +2674,38 @@ bgp_notification_print(netdissect_options *ndo,
...
+				/* a proper shutdown communication */
+				else if (shutdown_comm_length <= BGP_NOTIFY_MINOR_CEASE_ADMIN_SHUTDOWN_LEN) {
+					ND_TCHECK2(*tptr, shutdown_comm_length);
+					ND_PRINT((ndo, ", Shutdown Communication (length: %u): \"", shutdown_comm_length));
+					safeputs(ndo, tptr+1, shutdown_comm_length);
+					ND_PRINT((ndo, "\""));
+					remainder_offset += shutdown_comm_length + 1;
+				}
+				/* if there is trailing data, hexdump it */
+				if(length - (remainder_offset + BGP_NOTIFICATION_SIZE) > 0) {
```
Comment (Habbie) [inline print-bgp.c]: Use ND_TCHECK2 here as well?

**Comment by Habbie** [comment]
Looks like the included pcap was unparseable by recent libpcap (1.8.1 and master), I've replaced it with one that does work. My guess is this was not a code issue.

**Comment by job** [comment]
@infrastation is this good?

**Comment by infrastation** [comment]
I only had made a brief look, hopefully other team members will review as well in more detail.

**Comment by job** [comment]
ping, anyone?

**Comment by job** [comment]
ping anyone?

**Comment by infrastation** [comment]
Job, please give us some more time.

**Comment by job** [comment]
@infrastation sure thing! I appreciate your time and effort.

BTW, A new version of the draft was released. This version contains improved wording on what the protocol data unit looks like: https://tools.ietf.org/html/draft-ietf-idr-shutdown-03#section-2 Hopefully this clarifies what the patch does.

**Comment by job** [comment]
whoops, pressed the wrong button

**Comment by job** [comment]
any feedback? it has been 3 months :-)

**Comment by lukastribus** [comment]
I think you may get more feedback on the mailing list:
http://www.tcpdump.org/#mailing-lists

File: print-bgp.c [hunk 2]
```diff
@@ -259,11 +259,15 @@ static const struct tok bgp_notify_major_values[] = {
...
 /* draft-ietf-idr-cease-subcode-02 */
 #define BGP_NOTIFY_MINOR_CEASE_MAXPRFX  1
+/* draft-ietf-idr-shutdown-02 */
+#define BGP_NOTIFY_MINOR_CEASE_SHUT     2
+#define BGP_NOTIFY_MINOR_CEASE_RESET    4
+#define BGP_NOTIFY_MINOR_CEASE_ADMIN_SHUTDOWN_LEN   128
 static const struct tok bgp_notify_minor_cease_values[] = {
     { BGP_NOTIFY_MINOR_CEASE_MAXPRFX, "Maximum Number of Prefixes Reached"},
-    { 2,                        "Administratively Shutdown"},
+    { BGP_NOTIFY_MINOR_CEASE_SHUT,    "Administratively Shutdown"},
```
Comment (job) [inline print-bgp.c]: ack

File: print-bgp.c [hunk 3]
```diff
@@ -259,11 +259,15 @@ static const struct tok bgp_notify_major_values[] = {
...
     { 3,                        "Peer Unconfigured"},
-    { 4,                        "Administratively Reset"},
+    { BGP_NOTIFY_MINOR_CEASE_RESET,   "Administratively Reset"},
```
Comment (job) [inline print-bgp.c]: ack

File: tests/bgp-shutdown-communication.out [hunk 4]
```diff
@@ -0,0 +1,6 @@
+IP (tos 0xc0, ttl 255, id 40416, offset 0, flags [DF], proto TCP (6), length 126)
+    165.254.255.24.179 > 165.254.255.17.33202: Flags [P.], cksum 0xa924 (correct), seq 633078120:633078194, ack 1987858625, win 2172, options [nop,nop,TS val 921581890 ecr 3129863171], length 74: BGP
+	Notification Message (3), length: 74, Cease (6), subcode Administratively Shutdown (2), Shutdown Communication (length: 52): "This is a test of the shutdown communication system."
```
Comment (job) [inline tests/bgp-shutdown-communication.out]: ack

File: tests/bgp-shutdown-communication.out [hunk 5]
```diff
@@ -0,0 +1,6 @@
...
+IP (tos 0xc0, ttl 255, id 40416, offset 0, flags [DF], proto TCP (6), length 126)
+    165.254.255.24.179 > 165.254.255.17.33202: Flags [P.], cksum 0xa724 (correct), seq 0:74, ack 1, win 2172, options [nop,nop,TS val 921581890 ecr 3129863171], length 74: BGP
+	Notification Message (3), length: 74, Cease (6), subcode Administratively Reset (4), Shutdown Communication (length: 52): "This is a test of the shutdown communication system."
```
Comment (job) [inline tests/bgp-shutdown-communication.out]: ack

**Comment by job** [comment]
@fxlb many thanks for taking the time to review this PR. Would you might checking again? I implemented all changes suggested in your review.

**Comment by fxlb** [comment]
No need for '\' in the end of line 2682.
Updated and commited.
Thanks you !

**Pull request closed without merging**